### PR TITLE
[GOBBLIN-1302] add COMBINE_RETENTION_POLICIES config

### PR DESCRIPTION
**Pull request by arjun4084346**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!
@sv2000  please review

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1302

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
- adds a new config to specify combine retention policies
- adds aliases to some retention policy classes

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
added a unit test

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java [hunk 1]
```diff
@@ -204,7 +204,7 @@ static void waitJobCompletion(HelixManager helixManager, String workFlowName, St
       } else {
         // We have waited for WorkflowContext to get initialized,
         // so it is found null here, it must have been deleted in job cancellation process.
-        log.info("WorkflowContext not found. Job is probably cancelled.");
+        log.info("WorkflowContext not found. Job {} is probably cancelled.", jobName);
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java:207]: Seems unrelated to the change proposed in this PR. Can we remove this?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java [hunk 2]
```diff
@@ -62,10 +66,16 @@
  */
 public class CombineRetentionPolicy<T extends DatasetVersion> implements RetentionPolicy<T> {
 
+  public static final String COMBINE_RETENTION_POLICIES =
+      DatasetCleaner.CONFIGURATION_KEY_PREFIX + "combine.retention.policy.classes";
+  /**
+   * @Deprecated , use COMBINE_RETENTION_POLICIES instead.
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java:72]: +1.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java [hunk 3]
```diff
@@ -110,6 +105,30 @@ public CombineRetentionPolicy(Properties props) throws IOException {
...
+      retentionPolicyClasses = COMMA_BASED_SPLITTER.splitToList(props.getProperty(COMBINE_RETENTION_POLICIES));
+    } else {
+      retentionPolicyClasses = PropertiesUtils.getValuesAsList(props, Optional.of(RETENTION_POLICIES_PREFIX));
+    }
+
+    for (String retentionPolicyClass : retentionPolicyClasses) {
+      try {
+        builder.add((RetentionPolicy<T>) ConstructorUtils.invokeConstructor(
+            Class.forName(aliasResolver.resolve(retentionPolicyClass)), props));
+      } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException | InstantiationException
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java:123]: catch can be simplified to catch(ReflectiveOperationException).

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java [hunk 4]
```diff
@@ -110,6 +105,30 @@ public CombineRetentionPolicy(Properties props) throws IOException {
 
   }
 
+  private List<RetentionPolicy<T>> findRetentionPolicies(Properties props) {
+    List<String> retentionPolicyClasses;
+    ImmutableList.Builder<RetentionPolicy<T>> builder = ImmutableList.builder();
+    ClassAliasResolver<?> aliasResolver = new ClassAliasResolver<>(RetentionPolicy.class);
+
+    if (props.containsKey(COMBINE_RETENTION_POLICIES)) {
+      retentionPolicyClasses = COMMA_BASED_SPLITTER.splitToList(props.getProperty(COMBINE_RETENTION_POLICIES));
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java:114]: We should strip leading and trailing whitespaces using Splitter.on(',').trimResults().splitToList(...).

Comment (sv2000) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java:114]: Perhaps a good idea to do omitEmptyStrings() too.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/NewestKRetentionPolicy.java [hunk 5]
```diff
@@ -27,13 +27,15 @@
...
 
+import org.apache.gobblin.annotation.Alias;
 import org.apache.gobblin.data.management.retention.DatasetCleaner;
 import org.apache.gobblin.data.management.version.DatasetVersion;
 
 
 /**
  * Retains the newest k versions of the dataset.
  */
+@Alias("NewestK")
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/NewestKRetentionPolicy.java:38]: +1. Can we go all the way and add aliases to other retention policies too? There must be 3-4 more of them.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/PropertiesUtils.java [hunk 6]
```diff
@@ -84,6 +84,20 @@ public static long getPropAsLong(Properties properties, String key, long default
     return LIST_SPLITTER.splitToList(properties.getProperty(key));
   }
 
+  /**
+   * Extract all the values whose keys start with a <code>prefix</code>
+   * @param properties the given {@link Properties} instance
+   * @param prefix of keys to be extracted
+   * @return a list of values in the properties
+   */
+  public static List<String> getValuesAsList(Properties properties, Optional<String> prefix) {
```
Comment (sv2000) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/PropertiesUtils.java:93]: Can we add a unit test for this method?

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java (hunk 1 above)
Comment (arjun4084346) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java:207]: I added this in the description. Do you want me to create a separate PR for this?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java (hunk 4 above)
Comment (arjun4084346) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java:114]: It is there at Line 78

Comment (sv2000) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/policy/CombineRetentionPolicy.java:114]: Ack. Thanks!

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java (hunk 1 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixUtils.java:207]: Even though it is a trivial one-line change, I prefer to keep unrelated changes in separate PRs. The context of such unrelated changes is lost after some time and can be confusing to a new developer. My suggestion is to pull this change out of this PR.

**Review by sv2000 — CHANGES_REQUESTED** [review]
LGTM. Hope the Helix one-liner change can be removed from this PR.

**Review by sv2000 — APPROVED** [review]
+1. LGTM.
